src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/InMemoryEventStore.java

```diff
@@ -113,7 +143,8 @@ public void cleanupAfterSendingAttempt(boolean needRetry, long batchId) {
                 boolean result = eventBuffer.offerFirst(payloadToReinsert);
                 if (!result) {
                     LOGGER.error("Event buffer is full. Dropping newer payload to reinsert older payload");
-                    eventBuffer.removeLast();
+                    TrackerPayload removedPayload =  eventBuffer.removeLast();
+                    callback.onFailure(FailureType.TRACKER_STORAGE_FULL, false, Collections.singletonList(removedPayload));
```

Comment:
I commented about this above, but the more I think about it, the more it doesn't feel right for the InMemoryStorage to be given the responsibility for calling the EmitterCallback – I think the Emitter should fire the EmitterCallback for all the handled cases. If we distribute the calls in different places, it becomes more difficult to reason about when and how many times it is called. Also the storage layer can be replaced by users and the EventStore API does not provide them with the callback object in case they use a custom storage, so the storage full error won't be called in that case.

I think it would be better if the EventStore interface explicitly captured cases when the error occurs. Perhaps this function could return a list of payloads that could not be saved to the storage? Or we could make a dedicated callback for this specific error (not the EmitterCallback) and pass it as argument to the functions where it can occur (the Emitter would then listen to the callback and forward the errors to the EmitterCallback)?

Btw, it's awesome that you also capture this storage error in the EmitterCallback!

Reply:
The code would definitely be neater if only the Emitter has the EmitterCallback. I'll think about this callback idea